docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+```
+
+This model is potentially hugely expensive for the s3gw project
+because there isn't an implementation for the data replication with the SFS backend.
+
+## Active-Passive pipelines
+
+This model assumes N pipelines to be "allocated" on the cluster at the same time,
+but only one, the *active* pipeline, owns the *exclusive ownership*
+over the shared `PV`.
+
+Should the active pipeline suffer a failure, the next elected active pipeline,
+chosen to replace the old one, should pay a *time window* necessary to "transfer"
+the ownership over the shared `PV`.
+
+This means that, unlike the *Active-Active* solution, it is not guaranteed that
+the subsequent access made by a client after a failed one is always successful.
+
+It can happen that a client issues an access to the S3 service while the system is
+translating the passive pipeline into the active state; in this case, the issued
+access is expected to fail.
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+    ingress ==> radosgw2(radosgw)
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: red
+    linkStyle 2 stroke: red
+```
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw)
+    ingress ==> radosgw2(radosgw) == ownership transfer ==> PV((PV))
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: white
+    linkStyle 2 stroke: red
+```
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw)
+    ingress ==> radosgw2(radosgw) ==> PV((PV))
+
+    linkStyle 0 stroke: green
+    linkStyle 1 stroke: white
+    linkStyle 2 stroke: green
+    linkStyle 3 stroke: green
+```
+
+## One Active pipeline
+
+This model is logically equivalent to the *Active-Passive* model with the difference
+that only one pipeline is allocated on the cluster at a time.
+
+In the *One pipeline* model, the only notable perceived difference between
+the *Active-Passive* model are the worst performances obtained in the event of failure.
+
+In this scenario, the system should pay also the time needed to fully load a new
+pipeline, that unlike that *Active-Passive* model, is not preloaded.
+
+So, supposing that the new pipeline has scheduled to load on a node where the
+s3gw image is not cached, the system should pay the time needed to download
+the image from the registry before starting it.
+
+## Investigation's Rationale
+
+The 3 models described above have different performances and different implementation
+efforts. While the *Active-Active* model is expected to require a significant
+development effort due to its inherent complex nature, the *Passive-Active* model
+seems to offer a lower degree of implementation difficulty.
+
+In particular, the *One Active pipeline* model, is expected to work with nothing
+but the primitives normally available on any Kubernetes cluster.
+
+Given that the backend s3gw architecture is composed by:
```

Comment:
Should we add the ingress here? I am rather sure that we need to consider it in our model.

(Achieving an HA ingress may be out of scope, since that just kicks up the problem up one layer, but I suspect that while the ingress is functional, it can help us mask faults.)

Reply:
yes, it seems absolutely reasonable to add it to the model.